firestore/bulkwriter.go

```diff
@@ -0,0 +1,320 @@
+func (bw *BulkWriter) Update(doc *DocumentRef, updates []Update, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v; reason: %v", doc.ID, err)
+	}
+
+	if len(w) > 1 {
+		return nil, fmt.Errorf("firestore: too many writes sent to bulkwriter")
+	}
+
+	j := bw.write(w[0])
+	return j, nil
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	bw.openLock.Lock()
+	if !bw.isOpen {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+	bw.openLock.Unlock()
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) *BulkWriterJob {
+
+	j := &BulkWriterJob{
+		result: make(chan *pb.WriteResult, 1),
+		write:  w,
+		err:    make(chan error, 1),
+		ctx:    bw.ctx,
+	}
+
+	err := bw.bundler.Add(j, 0) // ignore operation size constraints; can't be inferred at compile time
+	if err != nil {
+		j.setError(err)
+	}
+	return j
+}
+
+// send transmits writes to the service and matches response results to job channels.
+// This method takes an empty interface to conform to the contract of the Bundler event handler.
+func (bw *BulkWriter) send(i interface{}) {
+	bwj := i.([]BulkWriterJob)
+
+	if len(bwj) == 0 {
+		return
+	}
+
+	var ws []*pb.Write
+	for _, w := range bwj {
+		ws = append(ws, w.write)
+	}
+
+	bwr := &pb.BatchWriteRequest{
+		Database: bw.database,
+		Writes:   ws,
+		Labels:   map[string]string{},
+	}
+
+	select {
+	case <-bw.ctx.Done():
+		return
+	default:
+		resp, err := bw.vc.BatchWrite(bw.ctx, bwr)
+		if err != nil {
+			// Do we need to be selective about what kind of errors we send?
+			for _, j := range bwj {
+				j.setError(err)
+			}
+			return
+		}
+		// Iterate over the response. Match successful requests with unsuccessful
+		// requests.
+		for i, res := range resp.WriteResults {
+			// Get the status code for this WriteResult
+			s := resp.Status[i]
+			c := s.GetCode()
+			if c != 0 { // Should we do an explicit check against rpc.Code enum?
+				j := bwj[i]
+				j.attempts++
+
+				// Do we need separate retry bundler
+				if j.attempts < maxRetryAttempts {
+					bw.bundler.Add(j, 0) // ignore operation size constraints; can't be inferred at compile time
```

Comment:
There's an error here that is checked above, but not here. Is that intentional?

Reply:
Yes. It's important to keep in mind that we have two related, but separate operations happening here:

+ Writes to the database, represented in memory by a `BulkWriterJob`
+ Requests to the service

The err check above fails when requests (which contain multiple writes) fail. The client returns an error object that we then pass back as the results to each individual write.

In this block, the request has succeeded (we got a response from the service), but an individual write has failed. In this case, we don't get an error object back from the client. Instead, we have to inspect the status code for each individual write and then either report a success, add a retry, or report a failure.

This code is missing the report a failure part, so I've added that.

Reply:
Are we talking about `bw.bundler.Add`? It returns an error that looking at the code is mostly unrelated. Perhaps this should just be:

``` golang
_ = bw.bundler.Add(j, 0) // ignoring the error because REASON
```